### PR TITLE
ARXIVNG-1414 Implement post-submission title check

### PR DESCRIPTION
**Pull request by erickpeirson**
In the classic system, we perform some QA checks within the submission UI itself. We need to preserve the behavior that these checks take place before/when the submission is finalized, but do this in a way that gives us flexibility and doesn't make requests take forever.

What I did here was to implement a hook mechanism on the Event class that allows us to bind functions for execution when an event instance is committed. 

This also implements the first pass at the submission worker process. This is a Celery worker that carries out asynchronous tasks. The basic approach is to use the ``arxiv.submission.tasks.is_async`` decorator, which registers the callback with the worker and sets up the machinery to dispatch an async task and persist its results.

From TD:

Registering event callbacks
===========================

The base :class:`Event` provides support for callbacks that are executed when
an event instance is committed. To attach a callback to an event type, use the
:func:`Event.bind` decorator. For example:

```python
   @SetTitle.bind()
   def do_this_when_a_title_is_set(event, before, after, agent):
       ...
       return []
```

Callbacks must have the signature ``(event: Event, before: Submission,
after: Submission, creator: Agent) -> Iterable[Event]``. ``event`` is the
event instance being committed that triggered the callback. ``before`` and
``after`` are the states of the submission before and after the event was
applied, respectively. ``agent`` is the agent responsible for any subsequent
events created by the callback, and should be used for that purpose.

The callback should not concern itself with persistence; that is handled by
:func:`Event.commit`. Any mutations of submission should be made by returning
the appropriate command/event instances.

The circumstances under which the callback is executed can be controlled by
passing a condition callable to the decorator. This should have the signature
``(event: Event, before: Submission, after: Submission, creator: Agent) ->
bool``; if it returns ``True``, the  callback will be executed. For example:

```python

   @SetTitle.bind(condition=lambda e, b, a, c: e.title == 'foo')
   def do_this_when_a_title_is_set_to_foo(event, before, after, agent):
       ...
       return []
```

### CODE REVIEW

File: core/arxiv/submission/rules/set_title.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+                'then,that,these,those,them,thus'.split(','))
+
+REMOVE_PUNCTUATION = str.maketrans(string.punctuation,
+                                   ' '*len(string.punctuation))
+"""Translator that converts punctuation characters into single spaces."""
+
+
+@SetTitle.bind(condition=lambda *a: not system_event(*a))
+@is_async
+def check_for_similar_titles(event: SetTitle, before: Submission,
```
Comment (erickpeirson) [inline core/arxiv/submission/rules/set_title.py]: Here's the routine for ARXIVNG-1414 itself

File: core/arxiv/submission/rules/set_title.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+def below_ascii_threshold(proportion: float) -> bool:
+    """Whether or not the proportion of ASCII characters is too low."""
+    threshold = get_application_config().get('TITLE_ASCII_THRESHOLD', 0.5)
+    return proportion < threshold
+
+
+def window() -> datetime:
+    """Get the time window for possible duplicate submissions."""
+    days = get_application_config().get('TITLE_SIMILARITY_WINDOW', 3*365/12)
```
Comment (JaimieMurdock) [inline core/arxiv/submission/rules/set_title.py]: Why not just use `365 / 4`?

File: Pipfile [hunk 3]
```diff
@@ -17,6 +17,10 @@ mysqlclient = "==1.3.13"
 mimesis = "==2.1.0"
 bleach = "==3.0.2"
 python-dateutil = "*"
+unidecode = "*"
+celery = "==4.1.0"
+arxiv-auth = "*"
+"c6a1273" = {path = "./core"}
```
Comment (JaimieMurdock) [inline Pipfile]: What is this syntax? Is the hash from the PIpfile.lock? Very curious to learn more.

File: core/arxiv/submission/rules/set_title.py (hunk 2 above)
Comment (erickpeirson) [reply core/arxiv/submission/rules/set_title.py]: Takes slightly less cognitive effort if we want to change it to 2 months or 4 months ;-)

Comment (erickpeirson) [reply core/arxiv/submission/rules/set_title.py]: "months"

File: Pipfile (hunk 3 above)
Comment (erickpeirson) [reply Pipfile]: Relative path; to install the package in`` ./core``

File: core/arxiv/submission/rules/set_title.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    Check for other submissions with very similar titles.
+
+    Query Submission (``arXiv_submissions`` table) for submissions with titles
+    that were created within the last 3 months.
+
+    Select previous matches for this submission id, and delete them -- from
+    SubmissionNearDuplicates (``arXiv_submission_near_duplicates`` table)
+
+    Get a Jaccard similarity indexer function
+    (arXiv::Submit::Jaccard::JaccardIndex->make_jaccard_indexer)
```
[the remark this answers is not in the document]
Comment (erickpeirson) [reply core/arxiv/submission/rules/set_title.py]: Whoops, yep that was left over from looking at the old code. Rewrote.

**Pull request merged**